[Test] 커뮤니티 서비스 로직에 대한 단위 테스트 코드 작성

eujjong18 opened the pull request:
## 🔎 Description
- 커뮤니티 API별 서비스 로직에 대한 단위 테스트 코드 작성
  - 후기 게시글 등록
  - 커뮤니티 홈 - 후기 목록 조회

## 💭 Issue
- closed #이슈번호

## 🗝 Key Changes
<!-- 주요 변경사항에 대해 작성해 주세요 -->
![image](https://github.com/user-attachments/assets/ff753f2f-4882-45b0-8e82-fde4cb7ed865)

## 🙏 To Reviewers
<!-- 리뷰어 전달사항을 작성해 주세요 -->
**❓ 질문)** 후기 게시글 등록 시, 이미지 저장 로직을 어떤 식으로 검증하는게 좋을지 의견을 여쭤보고 싶습니다!

저는 [이미지 저장] 메소드의 호출 여부로 이미지 저장 로직 실행 여부를 검증하고자 했는데요, 해당  메소드의 접근제어자가 private이다 보니 Mockito.verify( )를 통해 메소드 호출 여부를 검증하는 것이 불가능하더라구요. 

그래서 Reflection을 사용하여 saveImgList( ) 메소드를 강제로 실행하도록 하고, saveImgList( )에 대한 테스트 코드는 별도로 작성하지 않은 상태입니다.

그런데 구글링을 하다보니 private 메소드는 테스트를 지양하는 것이 좋고, private 메소드를 테스트해야 하는 상황이라면 재설계를 고려해볼 필요가 있다는 글을 보아서요
-> [참고자료 1](https://velog.io/@blackarea/Private-%EB%A9%94%EC%86%8C%EB%93%9C%EB%A5%BC-%ED%85%8C%EC%8A%A4%ED%8A%B8%ED%95%98%EA%B3%A0-%EC%8B%B6%EB%8B%A4%EA%B3%A0) / [참고자료 2](https://mangkyu.tistory.com/235)

혹시 이와 비슷한 상황에서 테스트를 진행한 경험이 있으시다면 어떤 식으로 처리하셨는지 궁금합니다 !

\+ 해당 부분의 코드는 아래를 참고해 주세요!

eujjong18 committed:
[test #74] 후기 게시글 등록 서비스 로직 테스트 코드 작성

eujjong18 committed:
[test #78] 후기 게시글 목록 조회 서비스 로직 테스트 코드 작성

eujjong18 commented on src/test/java/com/wellcome/WellcomeBE/domain/community/service/CommunityServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+        verify(communityRepository, times(1)).save(any(Community.class));
+
+        // - 이미지 저장 로직 실행 여부 검증
+        saveImgListMethod.invoke(communityService, Mockito.mock(Community.class), imgFileList);
```
Reflection을 사용하여 CommunityService의 saveImgList( ) 메소드(private)에 접근하도록 구현한 부분입니다

Hwater00 commented:
검증 포인트가 이 메소드가 호출되었는가 보다는 이미지가 정상적으로 저장되었는가라고 이해했습니다.  그래서Reflection을 사용하여 private 메소드를 강제 실행하는 것보다는 접근지정자를 protected 변경이나 아니면 prviate 메서드를 호출하는 테스트용 public 메서드 작성 후 이를 호출함을 권유드립니다.
다만, private으로 선언한 메소드가 객체 캡슐화 전략에 의한 잘못된 접근을 막고 있는 점이 망설여지는 요소지만, 저희가 도메인 기반으로 패키지 전략을 갖고 있어고 괜찮다고 생각해요. 한 번 확인해보고 시도해주세요!

eujjong18 committed:
[test #74] 이미지 등록 서비스 로직에 대한 테스트 코드 추가
- communityService의 saveImgList() 접근 제어자 private -> protected로 변경
- 이미지 저장 로직에 대한 테스트 코드 추가

eujjong18 commented:
> 검증 포인트가 이 메소드가 호출되었는가 보다는 이미지가 정상적으로 저장되었는가라고 이해했습니다. 그래서Reflection을 사용하여 private 메소드를 강제 실행하는 것보다는 접근지정자를 protected 변경이나 아니면 prviate 메서드를 호출하는 테스트용 public 메서드 작성 후 이를 호출함을 권유드립니다. 다만, private으로 선언한 메소드가 객체 캡슐화 전략에 의한 잘못된 접근을 막고 있는 점이 망설여지는 요소지만, 저희가 도메인 기반으로 패키지 전략을 갖고 있어고 괜찮다고 생각해요. 한 번 확인해보고 시도해주세요!

리뷰 감사합니다, 
회의 때 말씀해주신 내용 참고해서 해당 메소드의 접근제어자를 protected로 변경 후 테스트 재진행해보았습니다!
- 변경사항: 기존 '후기 게시글 등록' 테스트에서는 Reflection을 사용하지 않고 saveImgList( ) 호출 여부만 검증하도록 수정하고, 이미지 저장 로직에 대한 단위 테스트 코드를 추가했습니다.

Hwater00 commented on src/test/java/com/wellcome/WellcomeBE/domain/community/CommunityTestHelper.java:
CommunityTestHelper 클래스를 작성하니깐 중복 코드가 줄어들어 테스트 메서드에서 객체를 쉽게 생성할 수 있어 테스트 코드 작성에  유용해보입니다! 저도 참고해서 응용해보도록 하겠습니다.
